### PR TITLE
[ENH] - add support for Cohere assistants

### PR DESCRIPTION
**Pull request by smokestacklightnin**
This addresses Issue #227 

As always, comments and change requests are welcome and encouraged

### CODE REVIEW

File: ragna/assistants/_cohere.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        document_sources = [
+            {"title": source.id, "snippet": source.content} for source in sources
+        ]
+        return document_sources
+
+    async def _call_api(
+        self, prompt: str, sources: list[Source], *, max_new_tokens: int
+    ) -> AsyncIterator[str]:
+        response = await self._client.post(
+            "https://api.cohere.ai/v1/chat",
```
Comment (pmeier) [inline ragna/assistants/_cohere.py:39]: Cohere API is able to stream the response. This results in a better UX and we should always do that if the API supports it. From the [wording](https://docs.cohere.com/reference/chat), I guess they are using SSE to stream

> the response will be a JSON stream of events

If so, you can have a look at the OpenAI or Anthropic assistant on how to do that. If they for some reason mean that they actually respond with a JSON stream, have a look at the Google assistant on how to handle this.

Comment (smokestacklightnin) [reply ragna/assistants/_cohere.py:39]: It's a JSON stream. I'm looking now how to do it

**Comment by smokestacklightnin** [comment]
Quick status update

The code seems to receive the JSON stream http responses without error, but is not parsing them properly. No error is reported in the terminal console when running the UI, but the UI response says that something went wrong. I believe this has to do with the type of [streamedchatresponse](https://docs.cohere.com/reference/chat). I'll do more work to try to fix this.

Also, mypy complains when I run it after adapting [`AsyncIteratorReader`](https://github.com/Quansight/ragna/blob/8aa6bd09740be0bed6391fab00323caf4889f369/ragna/assistants/_google.py#L14-L25) from the google assistant code. I tried running mypy on that file too, and got similar complaints. Does anyone have any input?

**Comment by pmeier** [comment]
@smokestacklightnin I took the liberty and had a look to unblock you. I've pushed bfc868be8ecb590c90fe467daf2a62a7d00c466e that implements the streaming. While Cohere also uses a streaming HTTP request, in contrast to Google they don't stream JSON, but rather [JSONL](https://jsonlines.org/). Meaning, each line is valid JSON. Thus, we don't need an extra library to resolve the stream, but rather just ingest the stream line by line and convert to JSON ourselves.

**Comment by smokestacklightnin** [comment]
> @smokestacklightnin I took the liberty and had a look to unblock you. I've pushed [bfc868b](https://github.com/Quansight/ragna/commit/bfc868be8ecb590c90fe467daf2a62a7d00c466e) that implements the streaming. While Cohere also uses a streaming HTTP request, in contrast to Google they don't stream JSON, but rather [JSONL](https://jsonlines.org/). Meaning, each line is valid JSON. Thus, we don't need an extra library to resolve the stream, but rather just ingest the stream line by line and convert to JSON ourselves.

Thanks for the help! I'll continue working on this

**Comment by smokestacklightnin** [comment]
I added a `preamble_override`, which seems to be the last substantial change that is needed. I will then add links in the code and the FAQ documentation

File: docs/references/faq.md [hunk 2]
```diff
@@ -50,3 +50,14 @@
    ```bash
    export GOOGLE_API_KEY="XXXXX"
    ```
+
+### [Cohere](https://cohere.com/)
+
+1. > To use the API, you need an API key. You can create a key in the Cohere dashboard
+   > using your Cohere account
```
Comment (pmeier) [inline docs/references/faq.md:57]:
```suggestion
   > using your Cohere account.
```

File: ragna/assistants/_cohere.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    @classmethod
+    def display_name(cls) -> str:
+        return f"Cohere/{cls._MODEL}"
+
+    @property
+    def max_input_size(self) -> int:
+        return self._CONTEXT_SIZE
+
+    def _make_system_content(self) -> str:
```
Comment (pmeier) [inline ragna/assistants/_cohere.py:23]: Let's call this 

```suggestion
    def _make_preamble(self) -> str:
```

to align our terminology here with Cohere. We can't really align across assistants built into Ragna given that every provider uses something different.

File: ragna/assistants/_cohere.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        instruction = (
+            "You are a helpful assistant that answers user questions given the included context. "
+            "If you don't know the answer, just say so. Don't try to make up an answer. "
+            "Only use the included documents below to generate the answer."
+        )
+        return instruction
```
Comment (pmeier) [inline ragna/assistants/_cohere.py:29]: No need to store the string. Just return it.

```suggestion
        return (
            "You are a helpful assistant that answers user questions given the included context. "
            "If you don't know the answer, just say so. Don't try to make up an answer. "
            "Only use the included documents below to generate the answer."
        )
```

File: ragna/assistants/_cohere.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    def _make_source_documents(self, sources: list[Source]) -> list[dict[str, str]]:
+        document_sources = [
+            {"title": source.id, "snippet": source.content} for source in sources
+        ]
+        return document_sources
```
Comment (pmeier) [inline ragna/assistants/_cohere.py:35]: Same as above

```suggestion
        return [
            {"title": source.id, "snippet": source.content} for source in sources
        ]
```

**Review by pmeier — COMMENTED** [review]
Only a few minor comments left. Otherwise this is good to go. Thanks @smokestacklightnin!

File: docs/references/faq.md (hunk 2 above)
Comment (smokestacklightnin) [reply docs/references/faq.md]: Done

File: ragna/assistants/_cohere.py (hunk 3 above)
Comment (smokestacklightnin) [reply ragna/assistants/_cohere.py]: Done

File: ragna/assistants/_cohere.py (hunk 4 above)
Comment (smokestacklightnin) [reply ragna/assistants/_cohere.py]: Done

File: ragna/assistants/_cohere.py (hunk 5 above)
Comment (smokestacklightnin) [reply ragna/assistants/_cohere.py]: Done

**Comment by smokestacklightnin** [comment]
> Only a few minor comments left. Otherwise this is good to go. Thanks @smokestacklightnin!

Everything is done. Glad to contribute!

**Review by pmeier — APPROVED** [review]

**Pull request merged**